HW2/P3/mandelbrot.pyx

```diff
@@ -32,15 +34,67 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
-                z = 0
+        for i in prange(in_coords.shape[0], num_threads=2, schedule='static', chunksize=1):
+        #for i in range(in_coords.shape[0]):
+            for j in xrange(0, in_coords.shape[1], 8):
+                #c = in_coords[i, j]
+                c_real_avxval = AVX.make_float8((in_coords[i, j+7]).real,
+                                                (in_coords[i, j+6]).real,
+                                                (in_coords[i, j+5]).real,
+                                                (in_coords[i, j+4]).real,
+                                                (in_coords[i, j+3]).real,
+                                                (in_coords[i, j+2]).real,
+                                                (in_coords[i, j+1]).real,
+                                                (in_coords[i, j]).real)
+                c_imag_avxval = AVX.make_float8((in_coords[i, j+7]).imag,
+                                                (in_coords[i, j+6]).imag,
+                                                (in_coords[i, j+5]).imag,
+                                                (in_coords[i, j+4]).imag,
+                                                (in_coords[i, j+3]).imag,
+                                                (in_coords[i, j+2]).imag,
+                                                (in_coords[i, j+1]).imag,
+                                                (in_coords[i, j]).imag)
+                #z = 0
+                z_real_avxval = AVX.float_to_float8(0.0)
+                z_imag_avxval = AVX.float_to_float8(0.0)
+                all_one = AVX.float_to_float8(1.0)
+                counts_avxval = AVX.float_to_float8(0.0)
+                
+
                 for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
+                #    if magnitude_squared(z) > 4:
+                #        break
+                #    z = z * z + c
+                    
+                    magnitude_squared_z = AVX.add(AVX.mul(z_real_avxval, z_real_avxval), AVX.mul(z_imag_avxval, z_imag_avxval))
+                    mask = AVX.less_than(magnitude_squared_z, AVX.float_to_float8(4.0))
+                    
+                    if (AVX.signs(mask) == 0) :
                         break
-                    z = z * z + c
-                out_counts[i, j] = iter
+                    
+                    counts_avxval = AVX.add(counts_avxval, AVX.bitwise_and(mask, all_one))
+
+                    # compute z = z * z + c
+                    # if z = a + bi, c = x + yi
+                    # z * z = (a^2 - b^2) + (2ab)i
+                    # z_square_real = (a^2 - b^2)
+                    # z_square_imag = (2ab)
+                    z_square_real = AVX.sub(AVX.mul(z_real_avxval, z_real_avxval), AVX.mul(z_imag_avxval, z_imag_avxval))
+                    z_square_imag = AVX.add(AVX.mul(z_real_avxval, z_imag_avxval), AVX.mul(z_real_avxval, z_imag_avxval))
+                    
```

Comment:
Good use of comments, clearly explained how you derived z_square_real and z_square_imag!